packages/react-reconciler/src/ReactFiberHooks.new.js

```diff
@@ -395,6 +395,22 @@ export function renderWithHooks<Props, SecondArg>(
       current !== null && current.type !== workInProgress.type;
   }
 
+  // Reset the memoCache index and create a backup copy in case of a setState during render
+  // or error, either of which can leave the cache in an inconsistent state
+  const memoCache = workInProgress.memoCache;
+  let previousMemoCache = null;
+  if (memoCache !== null) {
+    previousMemoCache = memoCache.data.map(array => array.slice());
+    if (workInProgress.alternate != null) {
+      // Backup to the alternate fiber in case the component throws
+      workInProgress.alternate.memoCache = {
```

Comment:
We shouldn't need to mutate the `current` fiber at all. My understanding is we're starting with a copy-on-write approach, same as we do for fibers and hooks: inside `useMemoCache`, copy the cache from the previous current and put it on the work-in-progress fiber. If something errors, or the render is interrupted, the work-in-progress fiber will be dropped. So you don't need to do any resetting of the current cache or fiber.